src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -605,7 +613,11 @@ internal static bool Dispatch()
                         Debug.Assert(workItem is IThreadPoolWorkItem);
                         Unsafe.As<IThreadPoolWorkItem>(workItem).Execute();
                     }
-                    workItem = null;
+
+                    // Release refs and return to clean ExecutionContext and SynchronizationContext
+                    outerWorkItem = workItem = null;
+                    currentThread.ExecutionContext = null;
```

Comment:
> An alternative could be to run flow suppressed items in a Default context, rather than directly? Then the notifications would fire as it exited?

I think more generally, any change to execution context should coincide with sending notifications. Here too it should be checking for and sending notifications if necessary.

Reply:
And then the Dispatch-called methods won't have to do anything after invoking the work item.

Reply:
> The alternative would be to reset the context at the end of the Dispatch loop; fire any notifications, then reset again (though it wouldn't then fire for changes made in the notification handler, unless the reset was in a loop, but then it could be an infinite loop)

As you mentioned, since we're wanting to guarantee that thread pool threads are restored to a null EC, after sending change notifications, we could reset the EC again without sending change notifications. Changing the EC in a changed handler would be polluting anyway and in hindsight it may have been better generally to change the EC after sending change notifications, but I guess that would be more breaking.

Reply:
k will make that change